File: javascript/LeetCode/addTwoNumbers.js

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */

function ListNode(val, next) {
  this.val = (val===undefined ? 0 : val)
  this.next = (next===undefined ? null : next)
}

const sumCol = (val1, val2, carried) => {
    let total = val1 + val2 + carried
    let [result, carry] = total < 10 ? [total, 0] : [Number(total.toString().split('')[1]), 1]
    return [result, carry]
}

const addTwoNumbers = function(l1, l2) {
  const final = new ListNode(0, null)
  let [node, carried] = [final, 0]
  while (l1 || l2) {
      let num1 = l1 ? l1.val : 0
      let num2 = l2 ? l2.val : 0
      let colResult = sumCol(num1, num2, carried)
      node.val = colResult[0]
      carried = colResult[1]
      l1 = l1 ? l1.next : null
      l2 = l2 ? l2.next : null
      if (l1 || l2) {
        node.next = new ListNode(0, null)
        node = node.next
      }
      else {
          carried ? node.next = new ListNode(carried, null) : null
      }
  }
  return final
};

File: python/LeetCode/add_two_numbers.py


def sum_col(v1, v2, carried):
    total = v1 + v2 + carried
    result = total if total < 10 else int(str(total)[-1])
    carry = 0 if total < 10 else 1
    return result, carry

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):

    def addTwoNumbers(self, l1, l2):
        final = ListNode()
        node, carried = final, 0

        while l1 or l2:
            num1 = l1.val if l1 else 0
            num2 = l2.val if l2 else 0
            node.val, carried = sum_col(num1, num2, carried)
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
            if l1 or l2:
                node.next = ListNode()
                node = node.next
            else:
                node.next = ListNode() if carried else None
        return final

